src/plugins/dashboard/public/application/dashboard_app.tsx

```diff
@@ -17,75 +17,228 @@
+  redirectTo: DashboardRedirect;
+  embedSettings?: DashboardEmbedSettings;
 }
 
-export function initDashboardAppDirective(app: any, deps: RenderDeps) {
-  app.directive('dashboardApp', () => ({
-    restrict: 'E',
-    controllerAs: 'dashboardApp',
-    controller: (
-      $scope: DashboardAppScope,
-      $route: any,
-      $routeParams: {
-        id?: string;
-      },
-      kbnUrlStateStorage: IKbnUrlStateStorage,
-      history: History
-    ) =>
-      new DashboardAppController({
-        $route,
-        $scope,
-        $routeParams,
-        indexPatterns: deps.data.indexPatterns,
-        kbnUrlStateStorage,
-        history,
-        ...deps,
-      }),
-  }));
+export function DashboardApp({
+  savedDashboardId,
+  embedSettings,
+  redirectTo,
+  history,
+}: DashboardAppProps) {
+  const {
+    data,
+    core,
+    onAppLeave,
+    uiSettings,
+    indexPatterns: indexPatternService,
+    dashboardCapabilities,
+  } = useKibana<DashboardAppServices>().services;
+
+  const [lastReloadTime, setLastReloadTime] = useState(0);
+  const [lastContainerId, setLastContainerId] = useState<number>();
+  const [indexPatterns, setIndexPatterns] = useState<IndexPattern[]>([]);
+
+  const savedDashboard = useSavedDashboard(savedDashboardId, history);
+  const dashboardStateManager = useDashboardStateManager(savedDashboard, history);
+  const dashboardContainer = useDashboardContainer(dashboardStateManager, history, false);
+
+  const switchViewMode = useCallback(
+    (newMode: ViewMode) => {
+      dashboardStateManager?.switchViewMode(newMode);
+    },
+    [dashboardStateManager]
+  );
+
+  const refreshDashboardContainer = useCallback(
+    (lastReloadRequestTime?: number) => {
+      if (!dashboardContainer || !dashboardStateManager) {
+        return;
+      }
+      const changes = getChangesFromAppStateForContainerState({
+        dashboardContainer,
+        appStateDashboardInput: getDashboardContainerInput({
+          isEmbeddedExternally: Boolean(embedSettings),
+          dashboardStateManager,
+          lastReloadRequestTime,
+          dashboardCapabilities,
+          query: data.query,
+          switchViewMode,
+        }),
+      });
+      if (changes) {
+        if (getSearchSessionIdFromURL(history)) {
+          // going away from a background search results
+          removeQueryParam(history, DashboardConstants.SEARCH_SESSION_ID, true);
+        }
+
+        dashboardContainer.updateInput({
+          ...changes,
+          searchSessionId: data.search.session.start(),
+        });
+      }
+    },
+    [
+      history,
+      data.query,
+      embedSettings,
+      switchViewMode,
+      dashboardContainer,
+      data.search.session,
+      dashboardCapabilities,
+      dashboardStateManager,
+    ]
+  );
+
+  // Manage dashboard container subscriptions
+  useEffect(() => {
+    if (
+      !dashboardStateManager ||
+      !dashboardContainer ||
+      lastContainerId === dashboardContainer.runtimeId
```

Comment:
This isn't the most elegant solution, but it accomplishes what it's supposed to: Preventing duplicate change listeners being pushed to the dashboard state manager when the user saves the current dashboard as a new savedObject.

A better solution would be to ensure that the  `dashboardContainer` is null in this case, but I was unable to make that work correctly.

Reply:
Removed this workaround by fixing up the cleanup functions.